### PR TITLE
Optionally send PublishIntent additions/deletions to content store asynchronously

### PR DESCRIPTION
**Pull request by aldavidson**
Our [plan for safely swapping content-store over to a PostgreSQL database](https://docs.google.com/document/d/14B3YE978TDbujYnGHL0gm_gnf_EwD5MSHx7iRF24-90/edit#heading=h.wctys5xv7gdn) relies on us being able to suspend all writes/deletions for the length of time needed to run a full Mongo-to-Postgres export. This is currently around 90 minutes, but may be longer for the live content-store as it gets heavy sustained traffic.

When running this on the draft content-store, we found that some publish-intent writes were still going through even when the publishing-api workers were disabled. The numbers were low (less than 10 in 2hrs) but they still need to be handled.

We decided that the best option was to temporarily change publishing-api to send these updates to content-store asynchronously, for the duration of the switchover. (see the [Trello card](https://trello.com/c/a7ngC11v/952-figure-out-how-to-handle-observed-puts-of-publish-intents-while-publishing-api-workers-were-disabled) for full details). When the workers get switched back on at the end of the switchover, these updates will be pulled off the queue and sent to the content-store automatically.

We think this is safe, as publishing-api [already checks for uniqueness](https://github.com/alphagov/publishing-api/blob/d8ded888f56a2ae7909a8baf45968ed15e90423a/app/models/path_reservation.rb#L25-L30) of a path reservation before propagating to the content-store, so the risk of an out-of-order execution resulting in a problem should be very low.

Update 8th Dec 16:39:
I've successfully tested this with a scheduled publish on integration - 
image
...which showed up correctly in content-store:
```
> PublishIntent.order('created_at desc').first(1).pluck('base_path', 'created_at')
=> [["/government/news/test-new-news-story-5", Fri, 08 Dec 2023 16:32:50.131144000 UTC +00:00]]
```

So we'll merge this as close as possible to the start of the maintenance period on Monday 

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

This application is owned by the publishing platform team. Please let us know in #govuk-publishing-platform when you raise any PRs.

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -14,6 +14,7 @@
 /coverage
 /spec/reports
 spec/examples.txt
+.byebug_history
```
Comment (JonathanHallam) [inline .gitignore:17]: Should probably just be it's own commit

File: app/commands/put_publish_intent.rb [hunk 2]
```diff
@@ -5,7 +5,15 @@ def call
 
       if downstream
         payload = publish_intent
-        Adapters::ContentStore.put_publish_intent(base_path, payload)
+        enqueue = ENV.fetch("ENQUEUE_PUBLISH_INTENTS", false)
+        if enqueue == "true"
```
Comment (JonathanHallam) [inline app/commands/put_publish_intent.rb:9]: is there a reason this is stored as a string rather than a boolean?

File: app/commands/delete_publish_intent.rb [hunk 3]
```diff
@@ -2,7 +2,14 @@ module Commands
   class DeletePublishIntent < BaseCommand
```
Comment (JonathanHallam) [inline app/commands/delete_publish_intent.rb:2]: Can you add something in the commit message about why we're doing this please

File: spec/commands/put_publish_intent_spec.rb [hunk 4]
```diff
@@ -18,7 +18,7 @@
   context "when the downstream flag is set to false" do
     it "does not send any downstream requests" do
       expect(Adapters::ContentStore).not_to receive(:put_publish_intent)
-      
+
```
Comment (JonathanHallam) [inline spec/commands/put_publish_intent_spec.rb:21]: same as comment above about linting

**Review by JonathanHallam — CHANGES_REQUESTED** [review]
Few small questions/ tidy ups to be done but otherwise, good, thanks Al :)

File: .gitignore [hunk 5]
```diff
@@ -14,6 +14,7 @@
 /coverage
```
[the remark this answers is not in the document]
Comment (aldavidson) [reply .gitignore]: I've squashed and rebased, with a much more detailed multi-line-aware message

File: .gitignore (hunk 1 above)
Comment (aldavidson) [reply .gitignore]: yeah, I've removed this change & rebased.

File: app/commands/put_publish_intent.rb (hunk 2 above)
Comment (aldavidson) [reply app/commands/put_publish_intent.rb:9]: Yes - it's passed through as an environment variable from the Kubernetes Helm chart, which means it's ultimately going to passed through to the app as a string value:

```
$ RANDOM_ENV_VAR=true irb
irb(main):001> ENV['RANDOM_ENV_VAR']
=> "true"
```

File: spec/commands/put_publish_intent_spec.rb [hunk 6]
```diff
@@ -17,10 +17,35 @@
...
-
+      expect(Adapters::ContentStore).not_to receive(:put_publish_intent)
+      
       described_class.call(payload, downstream: false)
     end
   end
+
+  context "when the downstream flag is set to true" do
+    context "when the ENQUEUE_PUBLISH_INTENTS flag is present" do
+      context "and the ENQUEUE_PUBLISH_INTENTS flag is true" do
```
[the remark this answers is not in the document]
Comment (aldavidson) [reply spec/commands/put_publish_intent_spec.rb]: Fair point - I've removed the redundant line 27 (and its equivalent in `delete_publish_intent_spec.rb`) and added examples for when the flag is not set

File: app/commands/delete_publish_intent.rb (hunk 3 above)
Comment (aldavidson) [reply app/commands/delete_publish_intent.rb:2]: I've added a much more detailed explanation in the commit message.

File: app/commands/put_publish_intent.rb [hunk 7]
```diff
@@ -9,7 +9,7 @@ def call
         if enqueue == "true"
           PutPublishIntentWorker.perform_async(
             "base_path" => base_path,
-            "payload" => payload.to_json
+            "payload" => payload.to_json,
```
[the remark this answers is not in the document]
Comment (aldavidson) [reply app/commands/put_publish_intent.rb:12]: I rebased the two commits into one

File: app/workers/put_publish_intent_worker.rb [hunk 8]
```diff
@@ -5,12 +5,11 @@ class PutPublishIntentWorker
   include Sidekiq::Worker
   include PerformAsyncInQueue
 
-
   sidekiq_options queue: HIGH_QUEUE,
                   lock: :until_executing,
                   lock_args_method: :uniq_args,
                   on_conflict: :log
-  
+
```
[the remark this answers is not in the document]
Comment (aldavidson) [reply app/workers/put_publish_intent_worker.rb:12]: I rebased the two commits into one

File: spec/commands/delete_publish_intent_spec.rb [hunk 9]
```diff
@@ -9,17 +9,52 @@
...
+  context "when the downstream flag is set to false" do
+    it "does not send any downstream requests" do
+      expect(content_store).not_to receive(:delete_publish_intent)
 
       described_class.call(payload, downstream: false)
     end
   end
+
+  context "when the downstream flag is set to true" do
+    context "when the ENQUEUE_PUBLISH_INTENTS flag is present" do
```
[the remark this answers is not in the document]
Comment (aldavidson) [reply spec/commands/delete_publish_intent_spec.rb]: done - see my reply to your previous comment above on `put_publish_intent_spec.rb`

File: spec/commands/put_publish_intent_spec.rb (hunk 4 above)
Comment (aldavidson) [reply spec/commands/put_publish_intent_spec.rb:21]: Should be clearer now after rebasing

File: app/commands/put_publish_intent.rb (hunk 2 above)
Comment (JonathanHallam) [reply app/commands/put_publish_intent.rb:9]: totally agree

File: spec/commands/put_publish_intent_spec.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (JonathanHallam) [reply spec/commands/put_publish_intent_spec.rb]: I think something slightly funny has happened with the commits here because these changes are in the first one and then removed in the 2nd

[the remark this answers is not in the document]
Comment (aldavidson) [reply spec/commands/put_publish_intent_spec.rb]: I'll squash them both down to one commit before merging

[the remark this answers is not in the document]
Comment (aldavidson) [reply spec/commands/put_publish_intent_spec.rb]: done

**Review by JonathanHallam — APPROVED** [review]
thanks!
